### PR TITLE
Fix for remote installation tests

### PR DESCRIPTION
**Pull request by mloviska**
Local runs:

- SSH
Target http://dhcp228.suse.cz/tests/260
Controller http://dhcp228.suse.cz/tests/261
- VNC
Target http://dhcp228.suse.cz/tests/263
Controller http://dhcp228.suse.cz/tests/262

### CODE REVIEW

File: tests/remote/remote_controller.pm [hunk 1]
```diff
@@ -17,39 +17,51 @@ use testapi;
...
 use mm_network;
 use lockapi;
+use mmapi;
 
 sub run {
-    my ($self) = @_;
     my $target_ip;
+    my $children = get_children();
+    my $child_id = (keys %$children)[0];
+    my $password = 'nots3cr3t';
```
Comment (kravciak) [inline tests/remote/remote_controller.pm]: You can use type_password here - https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L1250

File: tests/remote/remote_target.pm [hunk 2]
```diff
@@ -1,34 +1,33 @@
...
 # Copying and distribution of this file, with or without modification,
 # are permitted in any medium without royalty provided the copyright
 # notice and this notice are preserved.  This file is offered as-is,
 # without any warranty.
 
-# Summary: Remote ssh & vnc installation
-# Maintainer: mkravec <mkravec@suse.com>
+# Summary: Controller/master for remote installations
+# Tags: poo#9576
+# Maintainer: Martin Kravec <mkravec@suse.com>
```
Comment (kravciak) [inline tests/remote/remote_target.pm]: You can change maintainer ;)

File: tests/remote/remote_target.pm [hunk 3]
```diff
@@ -1,34 +1,33 @@
...
 
 # poo#9576
 sub run {
-    # Notice MASTER system is ready for installation
-    assert_screen "remote_slave_ready", 200;
-    mutex_create "installation_ready";
-
-    # Wait until MASTER finishes installing the system
-    wait_for_children;
+    mutex_create("dhcp_client_configuration");
```
Comment (kravciak) [inline tests/remote/remote_target.pm]: Why do you need 2 mutexes here? - Why not just 'ssh_server_is_available'?
It seems that that controller job also waits for 2 consequent mutexes..

File: tests/remote/remote_target.pm [hunk 4]
```diff
@@ -1,34 +1,33 @@
...
+    assert_screen("remote_slave_ready", 350);
+    mutex_create("ssh_server_is_available");
+    assert_screen("grub2", 1200);
```
Comment (kravciak) [inline tests/remote/remote_target.pm]: Please add comment here because about why you wait 20 minutes

File: tests/remote/remote_controller.pm [hunk 5]
```diff
@@ -17,39 +17,51 @@ use testapi;
...
-    # Setup static NETWORK
-    configure_default_gateway;
-    configure_static_ip('10.0.2.12/15');
+    mutex_lock("dhcp_client_configuration", $child_id);
+    mutex_unlock("dhcp_client_configuration", $child_id);
+    mutex_lock("ssh_server_is_available", $child_id);
 
     # Wait until target becomes ready
-    mutex_lock "installation_ready";
+    # parse dhcpd.leases - now it should contain entries for all nodes
```
Comment (kravciak) [inline tests/remote/remote_controller.pm]: You can make it easier by setting EXTRABOOTPARAMS=hostname=XXX and then connecting to XXX.openqa.test - but that's up to you ;)

File: tests/remote/remote_controller.pm [hunk 6]
```diff
@@ -17,39 +17,51 @@ use testapi;
...
+        }
+        elsif ($line =~ /client-hostname\s+"(.*)"/) {
+            my $hostname = lc($1);
+            $dhcp_leases{$hostname} //= [];
+            push(@{$dhcp_leases{$hostname}}, $lease_ip);
+        }
+    }
+
+    mutex_unlock("ssh_server_is_available", $child_id);
+    select_console 'root-console';
```
Comment (kravciak) [inline tests/remote/remote_controller.pm]: As you said, you select user-console later, so this might not be needed

File: tests/remote/remote_target.pm (hunk 2 above)
Comment (mloviska) [reply tests/remote/remote_target.pm]: Maintainer field has been updated.

File: tests/remote/remote_target.pm (hunk 3 above)
Comment (mloviska) [reply tests/remote/remote_target.pm]: dhcp_client_configuration has been removed.

File: tests/remote/remote_target.pm (hunk 4 above)
Comment (mloviska) [reply tests/remote/remote_target.pm]: Remote target has to wait while whole installation process finishes and then it can continue with grub_test. Comment added in the remote_target.pm as well.

File: tests/remote/remote_controller.pm (hunk 6 above)
Comment (mloviska) [reply tests/remote/remote_controller.pm]: Agreed, root console is not mandatory.
http://dhcp228.suse.cz/tests/266#

File: tests/remote/remote_target.pm [hunk 7]
```diff
@@ -1,34 +1,34 @@
...
-    # Notice MASTER system is ready for installation
-    assert_screen "remote_slave_ready", 200;
-    mutex_create "installation_ready";
-
-    # Wait until MASTER finishes installing the system
-    wait_for_children;
+    assert_screen("remote_slave_ready", 350);
+    mutex_create("ssh_server_is_available");
+    # wait while whole installation process finishes
+    assert_screen("grub2", 1200);
```
Comment (okurz) [inline tests/remote/remote_target.pm]: This is something that should be improved. We have better methods for that, e.g. the [wait_boot](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/opensusebasetest.pm#L291) method. Please try to use that. It has a good documentation about the supported options.

File: tests/remote/remote_target.pm [hunk 8]
```diff
@@ -1,34 +1,34 @@
...
 # poo#9576
 sub run {
-    # Notice MASTER system is ready for installation
-    assert_screen "remote_slave_ready", 200;
-    mutex_create "installation_ready";
-
-    # Wait until MASTER finishes installing the system
-    wait_for_children;
+    assert_screen("remote_slave_ready", 350);
+    mutex_create("ssh_server_is_available");
```
Comment (okurz) [inline tests/remote/remote_target.pm]: Why did you change the name? I think "installation_ready" is a way better name because it also relates to VNC, "ssh_server" ready less so.

Comment (mloviska) [reply tests/remote/remote_target.pm]: Hello Oliver, during debugging I have used more than one mutex which eventually turned out to be redundant. Thank you for your remark! The name has been reverted to "installation_ready".

File: tests/remote/remote_target.pm (hunk 7 above)
Comment (mloviska) [reply tests/remote/remote_target.pm]: I really appreciate that you have introduced to me to wait_boot function. Thanks!
